tests/hardware/test_em27_webscraper.py

```diff
@@ -0,0 +1,148 @@
+
+
+@pytest.fixture
+def em27_scraper(qtbot) -> EM27Scraper:
+    """Fixture for EM27Scraper."""
+    return EM27Scraper()
+
+
+def test_init(subscribe_mock: MagicMock) -> None:
+    """Test EM27Scraper's constructor."""
+    scraper = EM27Scraper()
+    assert scraper._url == EM27_URL
+    subscribe_mock.assert_called_once_with(scraper.send_data, "em27.data.request")
+
+
+@patch("finesse.hardware.em27_scraper.get_em27sensor_data")
+def test_on_reply_received_no_error(
+    get_em27sensor_data_mock: Mock, em27_scraper: EM27Scraper, sendmsg_mock: Mock, qtbot
+) -> None:
+    """Test the _on_reply_received() method works when no error occurs."""
+    reply = MagicMock()
+    reply.error.return_value = QNetworkReply.NetworkError.NoError
+
+    # NB: This value is of the wrong type, but it doesn't matter here
+    get_em27sensor_data_mock.return_value = "EM27Properties"
+
+    # Check the correct pubsub message is sent
+    em27_scraper._on_reply_received(reply)
+    sendmsg_mock.assert_called_once_with("em27.data.response", data="EM27Properties")
+
+
+def test_on_reply_received_network_error(
+    em27_scraper: EM27Scraper, sendmsg_mock: Mock, qtbot
+) -> None:
+    """Test the _on_reply_received() method works when a network error occurs."""
+    reply = MagicMock()
+    reply.error.return_value = QNetworkReply.NetworkError.HostNotFoundError
+    reply.errorString.return_value = "Host not found"
+
+    # Check the correct pubsub message is sent
+    em27_scraper._on_reply_received(reply)
+    assert sendmsg_mock.call_args.args[0] == "em27.error"
+    assert isinstance(sendmsg_mock.call_args.kwargs["error"], EM27Error)
+    assert (
+        sendmsg_mock.call_args.kwargs["error"].args[0]
+        == "Network error: Host not found"
+    )
+
+
+@patch("finesse.hardware.em27_scraper.get_em27sensor_data")
+def test_on_reply_received_exception(
+    get_em27sensor_data_mock: Mock, em27_scraper: EM27Scraper, sendmsg_mock: Mock, qtbot
+) -> None:
+    """Test the _on_reply_received() method works when an exception is raised."""
+    reply = MagicMock()
+    reply.error.return_value = QNetworkReply.NetworkError.NoError
+
+    # Make get_em27sensor_data() raise an exception
+    error = Exception()
+    get_em27sensor_data_mock.side_effect = error
+
+    # Check the correct pubsub message is sent
+    em27_scraper._on_reply_received(reply)
+    sendmsg_mock.assert_called_with("em27.error", error=error)
+
+
+@patch("finesse.hardware.em27_scraper.QNetworkRequest")
+def test_send_data(
+    network_request_mock: Mock, em27_scraper: EM27Scraper, qtbot
+) -> None:  # adapted from test_opus_interface's test_request_command()
+    """Test EM27Scraper's send_data() method."""
+    request = MagicMock()
+    network_request_mock.return_value = request
+    reply = MagicMock()
+
+    with patch.object(em27_scraper, "_manager") as manager_mock:
+        with patch.object(em27_scraper, "_on_reply_received") as reply_received_mock:
```

Comment:
If you follow my suggestion of leaving `_on_reply_received()` alone:

```suggestion
        with patch("finesse.hardware.em27_scraper._on_reply_received") as reply_received_mock:
```